pkg/skaffold/runner/dev.go

```diff
@@ -34,35 +34,60 @@ import (
 var ErrorConfigurationChanged = errors.New("configuration changed")
 
 func (r *SkaffoldRunner) doDev(ctx context.Context, out io.Writer) error {
-	defer r.changeSet.reset()
-
 	r.logger.Mute()
 
-	if r.changeSet.needsAction() {
+	// lock, acquire the intents, and unlock
+	r.intentLock.Lock()
+	buildIntent, syncIntent, deployIntent := r.buildIntent, r.syncIntent, r.deployIntent
+	r.intentLock.Unlock()
+
+	if (r.changeSet.needsRedeploy && deployIntent) ||
+		(len(r.changeSet.needsRebuild) > 0 && buildIntent) ||
+		(len(r.changeSet.needsResync) > 0 && syncIntent) {
 		// if any action is going to be performed, reset the monitor's changed component tracker for debouncing
 		defer r.monitor.Reset()
 		defer r.listener.LogWatchToUser(out)
 	}
 
 	switch {
 	case r.changeSet.needsReload:
+		r.changeSet.resetReload()
 		return ErrorConfigurationChanged
-	case len(r.changeSet.needsResync) > 0:
+	case len(r.changeSet.needsResync) > 0 && syncIntent:
+		defer func() {
+			r.changeSet.resetSync()
+			r.resetSyncIntent()
```

Comment:
this is incorrect, this will wipe out any sync intent change that was "collected" while the devLoop is running. I would do a reset at the top for all three intents separately.

Reply:
I don't think I agree. say a user sends a sync request, then skaffold kicks off a sync. while the sync is happening, the user sends another sync request. do you think that should get "queued up" and cause another sync as soon as a syncable file has changed? my opinion is it shouldn't; any requests that get sent by the user **while the associated action is being performed** should be thrown away. to me this is a more predictable user experience than having the request result in an intent that is silently queued up.